HW3/P5/label_regions.cl

```diff
@@ -80,16 +87,35 @@ propagate_labels(__global __read_write int *labels,
     old_label = buffer[buf_y * buf_w + buf_x];
 
     // CODE FOR PARTS 2 and 4 HERE (part 4 will replace part 2)
+    if(idx_1D == 0){
+        for(int buf_xx = halo; buf_xx < buf_w - halo; buf_xx++){
+            for(int buf_yy = halo; buf_yy < buf_h - halo; buf_yy++){
+                if(buffer[buf_yy * buf_w + buf_xx] == last_index)
+                    buffer[buf_yy * buf_w + buf_xx] = last_label;
+                else if(buffer[buf_yy * buf_w + buf_xx] < w * h){
+                    last_index = buffer[buf_yy * buf_w + buf_xx];
+                    last_label = labels[last_index];
+                }
+            }
+        }
+    }
     
+    barrier(CLK_LOCAL_MEM_FENCE);
+
     // stay in bounds
     if ((x < w) && (y < h)) {
         // CODE FOR PART 1 HERE
         // We set new_label to the value of old_label, but you will need
         // to adjust this for correctness.
-        new_label = old_label;
+        old_label_1 = buffer[(buf_y - 1) * buf_w + buf_x];
+        old_label_2 = buffer[buf_y * buf_w + buf_x - 1];
+        old_label_3 = buffer[buf_y * buf_w + buf_x + 1];
+        old_label_4 = buffer[(buf_y + 1) * buf_w + buf_x];
+        if(old_label < w * h) new_label = min(min(min(min(old_label, old_label_1), old_label_2), old_label_3), old_label_4);
```

Comment:
After parts 2 and 4, you should use buffer[buf_w * buf_y + buf_x] instead of old_label.